bin/build

```diff
@@ -2,26 +2,50 @@
 
 set -e
 
-. bin/build_utils
+# Use Docker BuildKit to efficiently build the targets.
+# With BuildKit, only relevant layers for the target are built.
+export DOCKER_BUILDKIT=1
+DEBUG=false
 
-FULL_VERSION_TAG="$(full_version_tag)"
+while true ; do
+  case "$1" in
+    --debug ) DEBUG=true ; shift ;;
+     * ) if [ -z "$1" ]; then break; else echo "$1 is not a valid option"; exit 1; fi;;
+  esac
+done
 
-echo "---"
+# go to repo root folder for execution
+cd $(dirname $0)/..
 
-echo "Building secrets-provider-for-k8s Docker image"
+. bin/build_utils
 
-docker build \
-  --tag "secrets-provider-for-k8s:dev" \
-  --tag "secrets-provider-for-k8s:${FULL_VERSION_TAG}" \
-  --tag "secrets-provider-for-k8s:latest" \
-  --target "secrets-provider" \
-  .
+FULL_VERSION_TAG="$(full_version_tag)"
 
-echo "Building secrets-provider-for-k8s-redhat:$FULL_VERSION_TAG Docker image"
+echo "---"
 
-docker build --tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" \
-   --build-arg VERSION="$FULL_VERSION_TAG" \
-   --target "secrets-provider-for-k8s-redhat" \
-   .
+if [ $DEBUG = true ]; then
+  echo "Building secrets-provider-debug Docker image"
+
+  docker build \
+    --tag "secrets-provider-for-k8s-debug:latest" \
+    --target "secrets-provider-debug" \
+    .
+else
+  echo "Building secrets-provider-for-k8s:$FULL_VERSION_TAG Docker image"
+
+  docker build \
+    --tag "secrets-provider-for-k8s:dev" \
+    --tag "secrets-provider-for-k8s:${FULL_VERSION_TAG}" \
+    --tag "secrets-provider-for-k8s:latest" \
+    --target "secrets-provider" \
+    .
+
+  echo "Building secrets-provider-for-k8s-redhat:$FULL_VERSION_TAG Docker image"
+
+  docker build --tag "secrets-provider-for-k8s-redhat:${FULL_VERSION_TAG}" \
+     --build-arg VERSION="$FULL_VERSION_TAG" \
+     --target "secrets-provider-for-k8s-redhat" \
+     .
```

Comment:
While in general use of Dockerfile with the `target` is very obscure use, I can see why this approach is done here. For me though, this seems to be overall too much bundling of too many things into a single file, making it hard to read and making it even harder to maintain in the future. I think a better design might be
- A split-out of individual pieces:
```
- Dockerfile
- Dockerfile.debug (used the `FROM` linking to product tag
- Dockerfile.redhat
```
- Clear split of responsibilities within the `Dockerfile`:
```
<prerequisites>
<main container build out - single block>
<redhat container build out - single block>
<debug container build out - single block>
```

Reply:
Added it as a consideration for a following issue #150